### PR TITLE
[WIP] Using the robust solver for pyMBAR - avoiding convergence Failu…

### PR DESCRIPTION
**Pull request by RiesBen**
## Description
In this PR, we would like to propose switching the default solver, if pyMBAR > 4.0.0, such we have an improved convergence rate at the cost of minimal more time. -> less errors thrown. 

## Todos
- [ ] Implement feature / fix bug
- [ ] Add [tests](https://github.com/choderalab/openmmtools/tree/master/openmmtools/tests)
- [ ] Update [documentation](https://github.com/choderalab/openmmtools/tree/master/docs) as needed
- [ ] Update [changelog](https://github.com/choderalab/openmmtools/blob/master/docs/releasehistory.rst) to summarize changes in behavior, enhancements, and bugfixes implemented in this PR

## Status
- [ ] Ready to go

## Changelog message
```

```

### CODE REVIEW

**Comment by mikemhenry** [comment]
```
   if pymbar.version.short_version >= "4.0.0" and "solver_protocol" not in self._user_extra_analysis_kwargs:
AttributeError: module 'openmmtools.multistate.pymbar' has no attribute 'version'
```

**Comment by mikemhenry** [comment]
We should use this to compare versions https://packaging.pypa.io/en/latest/version.html#packaging.version.Version

**Comment by ijpulidos** [comment]
There are a few options here, from our meeting:

* Need to benchmark if this affects the performance (is it slower?)
* We would change the default behavior and communicate the performance impact in the CHANGELOG (release notes). Would be passed on construction of the analyzer.
* Realtime `analysis_kwargs` in https://github.com/choderalab/openmmtools/blob/cbff4c8391f56d62f4bd5f0813916cab4c2eea45/openmmtools/multistate/multistatesampler.py#L1540-L1564 could stay the same as it was before.
* Test with the overlap matrix from example in [pymbar issue thread](https://github.com/choderalab/pymbar/issues/419).
* Test with "production" simulations. Just to double check how this behaves with longer trajectories and more data.
* Make sure we have a CI matrix with pymbar 4 and the robust settings.

@mikemhenry @RiesBen please add any comments to this thread if I'm missing something. Thanks!

**Comment by mikemhenry** [comment]
> I'm not fully clued in on the discussion, but my suggestion would be, if possible, to keep the kwargs equal between realtime analysis and the final analysis. I.e. the failures are more likely to happen at low sample counts.

I think we decided against this since as the simulation goes on, the analysis takes longer so it would be better to do a fast method for the realtime analysis

**Comment by IAlibay** [comment]
> the analysis takes longer so it would be better to do a fast method for the realtime analysis

Assuming you're trying to avoid regressing, the "slow" method is the same method as pymbar 3. So you're not actually going "slower", indeed the "slow" method isn't actually clearly slower.

Going for the "fast" method probably increases your chances of getting errored values on fractional datasets.

**Comment by ijpulidos** [comment]
The real time analysis currently instantiates [its own `MultiStateSamplerAnalyzer`](https://github.com/choderalab/openmmtools/blob/b74a61a8df2b66d210c72ae22458efbc6bd0a188/openmmtools/multistate/multistatesampler.py#L1531) and it would always use the default user kwargs specified in the changes in this PR, since we don't really provide a way for the user to change those. That is, it would always be using the "robust" solver with PyMBAR 4.  I think this should be fine. I agree with @IAlibay on that with less samples we have more chances to fail, so "robust" here sounds like a good idea.

I agree that we are probably jumping the gun and maybe the "robust" solver is fast enough (maybe even faster than the pymbar 3 implementation we were using). This is something that we should probably double check to keep our sanity.

Other than that, I'd suggest that we probably want to adapt the test script in https://github.com/choderalab/pymbar/issues/419#issuecomment-1718386779 to make it a tests for our `MultiStateSamplerAnalyzer`, just to be sure we are doing things correctly, if that makes sense.

**Comment by IAlibay** [comment]
Maybe to add to this a little bit, what I'm advocating for is closer to:

1. For now, robust will do - we _shouldn't_ take a performance hit, because it's the same this as before.
2. In the future, optimizing things would be great.
3. Getting the "for now" solution out first would be great.

**Comment by mrshirts** [comment]
> In the future, optimizing things would be great.

I'm happy to meet with people to discuss.  "Robust" should be the best option, and should fail in relatively few cases. There's some interesting options of creating synthetic data to improve convergence, but that adds bias.

**Comment by mikemhenry** [comment]
[ala-thr.zip](https://github.com/user-attachments/files/17156868/ala-thr.zip)

**Comment by mikemhenry** [comment]
@IAlibay @ijpulidos ready for review!

**Review by IAlibay — COMMENTED** [review]
couple of comments otherwise lgtm

File: openmmtools/tests/test_sampling.py
Comment (mikemhenry) [reply openmmtools/tests/test_sampling.py:2619]: I thought about that but it felt like a lot to add for a single test, if we have more to download then we can add it.

File: openmmtools/tests/test_sampling.py
Comment (mikemhenry) [reply openmmtools/tests/test_sampling.py:2652]: I thought about that but we already do that in other tests, happy to add it. What threshold do we want to use to check that it is close?

Comment (IAlibay) [reply openmmtools/tests/test_sampling.py:2652]: This shouldn't be sampling anything, so ideally we should always be converging to a very similar value. If it changes, something changed enough to affect all our free energies. I would suggest something like 1e-4 or 1e-5 precision.

Comment (mikemhenry) [reply openmmtools/tests/test_sampling.py:2652]: I'll be honest, I forgot we were not sampling something, so I didn't even consider that we will get the same result each time we run (or really close to the same result)

File: openmmtools/tests/test_sampling.py
Comment (IAlibay) [inline openmmtools/tests/test_sampling.py:2656]: The fun way to do this is pytest.approx if you've not tried it before!

Comment (mikemhenry) [reply openmmtools/tests/test_sampling.py:2656]: I've used `numpy.isclose` but never `pytest.approx` before

Comment (IAlibay) [reply openmmtools/tests/test_sampling.py:2656]: It's the single value approximate, super fast and good error messages, have a go if you want!

**Comment by IAlibay** [comment]
I can't approve @mikemhenry 😅 , only comment - but lgtm

**Comment by mikemhenry** [comment]
Sounds good, I will wait for @ijpulidos to review it before we merge

File: openmmtools/tests/test_sampling.py
Comment (ijpulidos) [reply openmmtools/tests/test_sampling.py:2652]: We probably need a docstring here telling what the test is doing and how the nc file was generated. For future reference.

File: openmmtools/tests/test_sampling.py
Comment (ijpulidos) [inline openmmtools/tests/test_sampling.py:2660]: Thank you for adding the comments, this is now much more clear. Great work.

**Review by ijpulidos — APPROVED** [review]

**Comment by mikemhenry** [comment]
Oh I forgot to suggest a changelog entry, this is what I am thinking 

## New Behavior

* When using openmmtools with `pymbar 4` the `robust` solver will be used instead of the default solver for `pymbar 4`. To go back to the old behavior, pass in `analysis_kwargs={"solver_protocol": "default"}` when creating an analyzer. The behavior remains unchanged when using `pymbar 3`. See this issue https://github.com/choderalab/pymbar/issues/419 for more information.

**Comment by mikemhenry** [comment]
I am going to merge this in, but feel free to comment on the changelog entry!

**Pull request merged**
